src/WidgetBase.ts

```diff
@@ -160,6 +181,62 @@ export class WidgetBase<P extends WidgetProperties> extends Evented implements W
 				propertiesChangedListeners[i].call(this, evt);
 			}
 		}));
+
+		// Add a render decorator that will register callbacks for the v-node 'afterCreate' and
+		// 'afterUpdate' lifecycle methods to call the widget lifecycle methods onElementCreated
+		// and onElementUpdated
+		this.addDecorator('afterRender', (node: DNode) => {
+			// Cannot use @dojo/core/aspect because the vnode callback functions cannot change on each
+			// call to render.
+			decorate(node, (node: HNode) => {
+				const key = node.properties && node.properties.key;
+				if (key != null) {
+					this.setLifecycleProperties(node.properties, 'afterCreate');
+					node.properties.afterCreate = vnodeAfterCreate;
+				}
+			}, isHNode);
+
+			decorate(node, (node: HNode) => {
+				const key = node.properties && node.properties.key;
+				if (key != null) {
+					this.setLifecycleProperties(node.properties, 'afterUpdate');
+					node.properties.afterUpdate = vnodeAfterUpdate;
+				}
+			}, isHNode);
+			return node;
+		});
+	}
+
+	private setLifecycleProperties(properties: VirtualDomProperties, callbackKey: string) {
+		let lsProps = properties._lifecycleProperties;
+		if (!lsProps) {
+			properties._lifecycleProperties = lsProps = {
```

Comment:
I'm not sure if this is a good idea, isn't the widget reference going to end up on a dom node property? Which seems a bit leaky, and potentially cyclical? If we go down this route we'll now also be squatting on a property that we'd need to document - I don't think it'd be possible to actually stop it via interfaces which is a shame (due to the index signature).